### Constructors and Context teorie

Nase classa decimal tzn Decimal(x) muzeme zapsat vice zpusoby jako integer -- Decimal(10), string '0.1', tuples ale to je trochu zamotane. Floats radeji nepouzivat v podstate to zaeviduje s jejich nepresnosti.

- TUPLE CONSTRUCTOR
- 1.23 zapiseme jako +123 x 10 na -2
- -1.23  jako -123 x 10 na -2
- znamenko - cislice - exponent
- takze v tuplu se decimal zapisuje jako (sign(c1,c2...),exponent)
- stim ze s je 0 pokud mame kladne cislo
- s davame jako 1 pokud mame zaporne cislo
- -3.1415 zapiseme v tuple jako (1,(3,1,4,1,5),-4)
- exponent abychom posunuli carku v podstate o 4

Presnost constructu ovlivnuje matematicke operace ale construktor samotny neovlivnuje. 

- a = Decimal('0.12345'), b = Decimal('0.12345')
- DO KONTEXTU OBOJI ULOZI TAKTO S PRESNOSTI 5
- c = a+b => 0.25 ALE PRI VYPOCTU PRESNOST RESI

### Coding part

In [207]:
import decimal
from decimal import Decimal

In [210]:
Decimal(10), Decimal('10.1') #integer a string zapis

(Decimal('10'), Decimal('10.1'))

In [220]:
t = (0, (3,1,4,1,5), -4) #nejdriv musim takto
Decimal(t)
Decimal((0, (3,1,4,1,5), -4))
# pokud chci zapsat musim dalsi zavorky jinak netvorim tuple

Decimal('3.1415')

In [221]:
Decimal((1, (3,1,4,1,5), -4))#1ka pro zapornou hodnotu

Decimal('-3.1415')

In [222]:
format(0.1, '.25f')

'0.1000000000000000055511151'

In [223]:
Decimal(0.1) #nepouzivat float jako parametr pro Decimal

Decimal('0.1000000000000000055511151231257827021181583404541015625')

In [224]:
Decimal(0.1) == Decimal('0.1') #strig je ale funkcni

False

In [225]:
decimal.getcontext()

Context(prec=28, rounding=ROUND_HALF_EVEN, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])

In [228]:
decimal.DefaultContext.prec = 6
# nastavim presnost kontextu 

In [229]:
a = Decimal('0.123456789')

In [230]:
a #ale i pres prenost 6 to vidi spravne

Decimal('0.123456789')

In [232]:
decimal.DefaultContext.prec = 2
a = Decimal('0.12345')
b = Decimal('0.12345')
c = a+b
print(a,b,c)
#Ale zde vidime ze pro vypocty se presnost meni
# tzn secte a zaokrouhli 

0.12345 0.12345 0.25


In [235]:
decimal.DefaultContext.prec = 6
a = Decimal('0.12345')
b = Decimal('0.12345')
print(a+b)
with decimal.localcontext() as ctx:
    ctx.prec =2
    c = a + b
    print(f'{c} with local context')
print(f'{c} with global context')

0.24690
0.25 with local context
0.25 with global context


Nejdrive jsme pocitali s presnosti 6 a tudiz mame 0.24690 ale pote doslo k prechodu na presnost 2. Pokud to ale znovu neprepisi stale si pamatuje hodnotz z local contextu.